### PR TITLE
Feature/mscasso/sp 3430 add line headers filter

### PR DESCRIPTION
**Pull request by mscasso-scanoss**
## Summary by CodeRabbit

* **New Features**
  * Added --skip-headers flag for fingerprint and scan commands to exclude license headers, comments, and imports at file starts

* **Tests**
  * Added comprehensive tests verifying header-skipping behavior across many languages, edge cases, and binary/unsupported files

* **Chores**
  * Bumped package version to 1.42.0

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: src/scanoss/line_filter.py [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+        try:
+            # Decode content to UTF-8
+            text_content = contents.decode('utf-8')
+        except UnicodeDecodeError:
+            # If decoding fails, return original content
+            self.print_debug(f'Skipping line filter due to UTF-8 decode error: {file}')
+            return contents, 0
+
+        # Detect language
+        language = self.detect_language(file)
```
Comment (agustingroh) [inline src/scanoss/line_filter.py]: should we move detect_language before contents.decode to avoid decode of unsupported languages?

File: src/scanoss/header_filter.py [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+            return contents, 0
+
+        # Detect language
+        language = self.detect_language(file)
+
+        # If language is not supported, return original content
+        if not language:
+            self.print_debug(f'Skipping line filter for unsupported language: {file}')
+            return contents, 0
+
```
Comment (eeisegn) [inline src/scanoss/header_filter.py]: remove blank line

File: src/scanoss/header_filter.py [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+        Parameters
+        ----------
+            max_lines: int
+                Maximum number of lines to return (None = unlimited by default)
+        """
+        super().__init__(debug, trace, quiet)
+        self.patterns = LanguagePatterns()
+        self.max_lines = max_lines
+
+    def filter(self, file: str, bin_file: bool, contents: bytes) -> tuple:
```
Comment (eeisegn) [inline src/scanoss/header_filter.py]: missing tuple type spec

File: src/scanoss/header_filter.py [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+        # Split into lines keeping line endings
+        lines = text_content.splitlines(keepends=True)
+        total_lines = len(lines)
+        self.print_debug(f'Analyzing {total_lines} lines for file: {file}')
+
+        # Find first implementation line (optimized - stops at first match)
+        implementation_start = self.find_first_implementation_line(lines, language)
+
+        # If no implementation, return empty
+        if implementation_start is None or implementation_start > len(lines):
```
Comment (eeisegn) [inline src/scanoss/header_filter.py]: Should this be `>=` or will it always be `>`?

File: src/scanoss/header_filter.py [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+            self.print_debug(f'Filtered out {line_offset} lines from beginning of {file} (language: {language})')
+
+        # Get lines from implementation (0-indexed)
+        start_idx = implementation_start - 1
+        filtered_lines = lines[start_idx:]
+
+        # Apply line limit if configured
+        if self.max_lines is not None:
+            original_count = len(filtered_lines)
+            filtered_lines = filtered_lines[:self.max_lines]
```
Comment (eeisegn) [inline src/scanoss/header_filter.py]: need error check if `max_lines` is longer than the filtered content.

File: src/scanoss/header_filter.py [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+            Tuple of (filtered_content: bytes, line_offset: int)
+            - filtered_content: Filtered content from where implementation code begins
+            - line_offset: Number of lines skipped from the beginning (0 if no filtering)
+        """
+        self.print_debug(f'HeaderFilter processing file: {file}')
+
+        # If binary file, return without processing
+        if bin_file:
+            self.print_debug(f'Skipping line filter for binary file: {file}')
+            return contents, 0
```
Comment (eeisegn) [inline src/scanoss/header_filter.py]: should be `b'''`?

File: src/scanoss/header_filter.py [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+        # Join lines and convert to bytes
+        filtered_content = ''.join(filtered_lines)
+        return filtered_content.encode('utf-8'), line_offset
+
+
+    def detect_language(self, file_path: str) -> Optional[str]:
+        """Detect language based on file extension"""
+        path = Path(file_path)
+        extension = path.suffix.lower()
+
```
Comment (eeisegn) [inline src/scanoss/header_filter.py]: remove blank line or add comment

**Review by eeisegn — APPROVED** [review]
